opsdroid/database/matrix/__init__.py

```diff
@@ -0,0 +1,249 @@
+            )
+            return
+        ori_data = ori_data.content
+
+        if self._single_state_key:
+            value = {key: value}
+
+        elif not isinstance(value, dict):
+            raise ValueError(
+                "When the matrix database is configured with single_state_key=False, "
+                "the value passed must be a dict."
+            )
+
+        sub_key = list(value.keys())[0]
+
+        if list(value.keys())[0] in ori_data:
+            if self._single_state_key:
+                existing_val = await self.get(sub_key)
+            else:
+                existing_val = await self.get({key: sub_key})
+            if existing_val == value[sub_key]:
+                _LOGGER.debug("Not updating matrix state, as content hasn't changed.")
+                return
+
+        if self.should_encrypt:
+            room_event = await self.opsdroid.send(
+                GenericMatrixRoomEvent(content=value, event_type=self._event_type)
+            )
+            data = {**ori_data, sub_key: {"encrypted_val": room_event.event_id}}
+        else:
+            data = {**ori_data, **value}
+
+        _LOGGER.debug(f"Putting {key} into matrix room {self.room_id} with {data}")
+
+        await self.opsdroid.send(
+            MatrixStateEvent(
+                self._event_type,
+                content=data,
+                target=self.room_id,
+                connector=self.connector,
+                state_key=state_key,
+            )
+        )
+
+        return True
+
+    async def get(self, key):
+        """Get a value from the database for a given key."""
+
+        if self.should_migrate:
+            await self.migrate_database()
+
+        if not self._single_state_key and not isinstance(key, dict):
+            _LOGGER.error(
+                "When the matrix database is configured with single_state_key=False, key must be a dict."
+            )
+
+        # If the single state key flag is set then use that else use state key.
+        state_key = (
+            "" if self._single_state_key is True else self._single_state_key or key
+        )
+
+        if isinstance(state_key, dict):
+            state_key, key = list(state_key.items())[0]
+
+        _LOGGER.debug(
+            f"Getting {key} from matrix room {self.room_id} with state_key={state_key}."
+        )
+
+        data = await self.connector.connection.room_get_state_event(
+            room_id=self.room_id, event_type=self._event_type, state_key=state_key,
+        )
+        if isinstance(data, RoomGetStateEventError):
+            _LOGGER.error(
+                f"Error getting {key} from matrix room {self.room_id}: {data.message}({data.status_code})"
+            )
+            return
+        data = data.content
+
+        _LOGGER.debug(f"Got {data} from state request.")
+
+        for k, v in data.items():
+            if isinstance(v, dict) and "encrypted_val" in v:
+                resp = await self.connector.connection.room_get_event(
```

Comment:
We probably need to at least log something here if we fail to decrypt this event because of missing keys. (i.e. it was sent a long time ago with a different instance of opsdroid that can no longer decrypt it etc)